### PR TITLE
Docs: Create Sparse Merkle Tree test spec

### PR DESCRIPTION
**Pull request by bvrooman**
Steps toward an automated test suite:
1. Define the test spec (this PR)
2. Codify test spec in a machine readable format: https://github.com/FuelLabs/fuel-merkle/issues/67
3. Read machine readable test spec and generate test suite: https://github.com/FuelLabs/fuel-merkle/issues/68

### CODE REVIEW

**Comment by Voxelot** [comment]
This kind of works, although I was hoping we'd have something more along the lines of a machine readable format e.g. toml or json that the library could execute against.

**Comment by bvrooman** [comment]
> This kind of works, although I was hoping we'd have something more along the lines of a machine readable format e.g. toml or json that the library could execute against.

I considered a DSL approach for this. I think that such an approach could be built alongside this spec.

Regardless of the ultimate testing strategy implemented by an SMT library, we would some sort of specification document, since any sort of machine readable test spec still has to be manually compiled.

**Comment by Voxelot** [comment]
Agree that we shouldn't gate on the automated harness as long we we have this spec manually implemented for now. Please create an issue for the machine readable spec if you haven't already and we can discuss that in more detail later.

File: docs/test-specs/sparse_merkle_tree_tests.md [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+
+Tests the default root given no update or delete operations.
+
+**Inputs**:
+
+_No inputs_
+
+**Outputs**:
+
+- The expected root signature: `0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855`
```
Comment (Voxelot) [inline docs/test-specs/sparse_merkle_tree_tests.md:52]: Should we mark this part of the spec as WIP until the corresponding issue in celestia (https://github.com/celestiaorg/smt/issues/67) is resolved?

File: docs/test-specs/sparse_merkle_tree_tests.md [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+Tests the root after performing one update call with null data. Updating the empty tree with null data does not change the root, and the expected root remains the default root. This test expects a root signature identical to that produced by [Test Empty Root](#test-empty-root).
+
+**Inputs**:
+
+1. Update the empty tree with leaf key `0u32` (4 bytes, big endian) and null leaf data `"\0"` (1 byte)
+
+**Outputs**:
+
+- The expected root signature: `0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855`
```
Comment (Voxelot) [inline docs/test-specs/sparse_merkle_tree_tests.md:235]: seems like this is also affected by the celestia issue

**Comment by bvrooman** [comment]
> I agree that we shouldn't gate on the automated harness as long we we have this spec manually implemented for now. Please create an issue for the machine readable spec if you haven't already and we can discuss that in more detail later.

I wrote up two tickets for automated harness:
1. https://github.com/FuelLabs/fuel-merkle/issues/67
2. https://github.com/FuelLabs/fuel-merkle/issues/68

I haven't fully implemented this spec as unit tests in the SMT PR yet (as you saw, a couple tests are missing or deviate in some way). I would like to make sure we have a consensus on this document before I codify it in the library. A few tests are affected by the Celestia spec discrepancy, so there's some ambiguity there. As you suggest, I may just write that those tests are still WIP until the issue is resolved.

File: docs/test-specs/sparse_merkle_tree_tests.md (hunk 1 above)
Comment (bvrooman) [reply docs/test-specs/sparse_merkle_tree_tests.md:52]: Yes, I think that's a good interim solution. I will add that affected tests are WIP and reference the issue in this document. Once we've determined the correct root, I will update those sections, and bump the document version.

**Review by adlerjohn — COMMENTED** [review]
Hmm. Shouldn't keys be 32 bytes?

**Comment by bvrooman** [comment]
> Hmm. Shouldn't keys be 32 bytes?

Key input can be any length - upon calling `update(&key, &data)`, the key input gets hashed into 32 bytes (and similarly, the data input gets hashed into 32 bytes). These hashed keys and data are what become leaves and nodes. The intuition is that `update(..)` can be called with any key, as long as the key can be represented as an array of bytes, and we don't expose the hashing keys to the end user.

**Comment by bvrooman** [comment]
Side node: I am changing the empty root value to the zero sum (`000...`) in the meantime, until we can come to a conclusion on the issue in Celestia. I am going with the zero sum for now since that is currently what both code implementations produce, so we are at least compliant (possibly wrong, but at least compliant!).

**Comment by Voxelot** [comment]
Should we also make a repo of the Go code used to generate these roots from celestia's impl as a future reference in case we want to update these tests?

**Review by Voxelot — APPROVED** [review]

**Comment by bvrooman** [comment]
> Should we also make a repo of the Go code used to generate these roots from celestia's impl as a future reference in case we want to update these tests? Or just put them in a local subdir somewhere in the docs of this repo?

I would suggest that we fork [Celestia's SMT repo](https://github.com/celestiaorg/smt). In the fork, I can put up the tests I wrote to generate the roots for this document. We could then add or modify the tests in the fork and record the outputs of those tests in this document. I imagine with a few iterations, we could get a high degree of confidence in our test coverage.

**Comment by Voxelot** [comment]
That makes sense, from the fork it would be easy to PR the tests back to the official repo once we're happy with the coverage.

**Comment by adlerjohn** [comment]
I'm okay with forking the Go repo only if it's used purely to upstream the changes. We don't want to maintain a Go implementation of the SMT. With that said, do we really need to fork the repo, or can @bvrooman just contribute the change upstream immediately?

**Comment by adlerjohn** [comment]
> the key input gets hashed into 32 bytes

Wait no, the key is not hashed. Or rather, should not be. See https://github.com/celestiaorg/smt/issues/40. Note that the SMT spec _does not_ require hashed keys. That's left to the application.

**Comment by bvrooman** [comment]
> > the key input gets hashed into 32 bytes
> 
> Wait no, the key is not hashed. Or rather, should not be. See [celestiaorg/smt#40](https://github.com/celestiaorg/smt/issues/40). Note that the SMT spec _does not_ require hashed keys. That's left to the application.

I see! On the Fuel side, I can modify the SMT interface to take in a `Bytes32`. In order for the tests to line up with the Celestia results, I can continue to hash the keys via sha256, and pass in the resulting 32. I see that there is already a [PR in Celestia ](https://github.com/celestiaorg/smt/pull/64) to move towards raw keys. 

For the test vectors in this spec, I will move the hashing from the library side to the client side to produce 32 byte keys. If we specify that the test inputs are generated by using Sha256 on an input to create the 32 byte keys, we can ensure continuity of keys between implementations.

**Comment by bvrooman** [comment]
> I'm okay with forking the Go repo only if it's used purely to upstream the changes. We don't want to maintain a Go implementation of the SMT. With that said, do we really need to fork the repo, or can @bvrooman just contribute the change upstream immediately?

The fork may not be necessary - it was just an idea to stage changes coming from us before committing to Celestia. I anticipate that the test spec will undergo at least couple of iterations. But I'm also happy to just commit any changes directly to Celestia. 

It looks like the migration to using raw keys definitely affects the test cases I used to generate the roots for this document. I can put that up in a Celestia PR, but it may be worth waiting for that change to go in before I commit it anyway.

**Comment by adlerjohn** [comment]
The Celestia SMT is going slow because it's not on the required path for now. I would say ignore adding anything to that repo for the time being.

**Review by Voxelot — APPROVED** [review]

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**